### PR TITLE
added buffer importer

### PR DESCRIPTION
**Pull request by nginnever**
basically this is the same as importing from a file but under the assumption that the importer should take a prebuffered file in memory up to the js max of 1gb. 

also replaced the out of bounds parameters in fixedSizeChunker for chunks larger than the fixed size. 

also added condition at the top of tests to look for empty dir needed that github didn't push for testing nested dirs

### CODE REVIEW

File: src/chunker-fixed-size.js [hunk 1]
```diff
@@ -23,7 +23,7 @@ function FixedSizeChunker (size) {
       var chunk = new Buffer(size, 'binary')
       var newBuf = new Buffer(buf.length - size, 'binary')
       buf.copy(chunk, 0, 0, size)
-      buf.copy(newBuf, 0, size - 1, buf.length - size)
+      buf.copy(newBuf, 0, size, buf.length)
```
Comment (diasdavid) [inline src/chunker-fixed-size.js]: thank you for catching this one! :D

File: src/index.js [hunk 2]
```diff
@@ -15,28 +15,45 @@ const CHUNK_SIZE = 262144
...
+    if(!Buffer.isBuffer(options.buffer)) { 
+      return callback(new Error('buffer importer must take a buffer'))
+    }
+    bufferImporter(options.buffer, callback)
+  } else if(options.stream) {
+    if(!(typeof options.stream.on === 'function')) {
+      return callback(new Error('stream importer must take a readable stream'))
+    }
+    streamImporter(options.stream, callback)
+  }
```
Comment (diasdavid) [inline src/index.js]: there should be a `return` statement here somewhere, so that if someone passes a path and a buffer, it doesn't fire two importers

File: src/index.js [hunk 3]
```diff
@@ -15,28 +15,45 @@ const CHUNK_SIZE = 262144
...
 
   function fileImporter (path, callback) {
     const stats = fs.statSync(path)
     if (stats.size > CHUNK_SIZE) {
       const links = [] // { Hash: , Size: , Name: }
-
       fs.createReadStream(path)
         .pipe(new FixedSizeChunker(CHUNK_SIZE))
         .pipe(through2((chunk, enc, cb) => {
+
```
Comment (diasdavid) [inline src/index.js]: didn't standard complained about starting a new code block with an empty line?

File: src/index.js [hunk 4]
```diff
@@ -166,9 +182,78 @@ exports.import = (options, callback) => {
...
       })
   }
+  function bufferImporter (buffer, callback) { 
+    const links = [] // { Hash: , Size: , Name: }
+    if (buffer.length > CHUNK_SIZE) {
+      var fsc = new FixedSizeChunker(CHUNK_SIZE)
+      fsc.write(buffer)
+      fsc.end()
+      fsc.pipe(through2((chunk, enc, cb) => {
+
```
Comment (diasdavid) [inline src/index.js]: same here?

File: tests/test-import.js [hunk 5]
```diff
@@ -18,6 +18,24 @@ describe('layout: importer', function () {
   const dirBig = __dirname + '/test-data/dir-big'
   const dirNested = __dirname + '/test-data/dir-nested'
 
+  // check to see if missing empty dirs need to be created
+  fs.exists(__dirname+'/test-data/dir-nested/dir-another', function(exists){
+    if(!exists){
+      fs.mkdir(__dirname+'/test-data/dir-nested/dir-another', function(){
+        fs.exists(__dirname+'/test-data/dir-nested/level-1/level-2', function(exists){
```
Comment (diasdavid) [inline tests/test-import.js]: this is not dependent of creating `dir-another`. You can just leave leave lines 32-35 to do it

File: tests/test-import.js [hunk 6]
```diff
@@ -18,6 +18,24 @@ describe('layout: importer', function () {
...
+          }
+        })
+      })
+    }
+    fs.exists(__dirname+'/test-data/dir-nested/level-1/level-2', function(exists){
+      if(!exists){
+        fs.mkdir(__dirname+'/test-data/dir-nested/level-1/level-2')
+      }
+    })
+  })
```
Comment (diasdavid) [inline tests/test-import.js]: :+1:

**Comment by diasdavid** [comment]
LGTM (other than the small things) , thank you @nginnever :)

If you can, let's get this module prepared to run in the browser next (standard webpack + karma testing setup)

File: src/index.js (hunk 2 above)
Comment (nginnever) [inline src/index.js]: yup, was actually thinking about that yesterday, on it!

File: src/index.js (hunk 3 above)
Comment (nginnever) [inline src/index.js]: oops ```"pre-commit" : ["lint"]``` isn't in the package.json, good catch.

File: tests/test-import.js (hunk 5 above)
Comment (nginnever) [inline tests/test-import.js]: been playing with this, i can't get the multihash to be the same unless ```dir-another``` exists. Are you sure we don't need to check for and create this folder? Or am I missing something.

Comment (diasdavid) [inline tests/test-import.js]: I mean that this:

```
fs.exists(__dirname+'/test-data/dir-nested/dir-another', function (exists) {
  if (!exists) {
    fs.mkdir(__dirname+'/test-data/dir-nested/dir-another', function () {})
  }
  fs.exists(__dirname+'/test-data/dir-nested/level-1/level-2', function(exists){
    if (!exists) {
      fs.mkdir(__dirname+'/test-data/dir-nested/level-1/level-2')
    }
  })
})
```

should be enough (you were checking and creating the level-2 folder twice)

Comment (diasdavid) [inline tests/test-import.js]: also note that fs.exists is deprecated:
https://nodejs.org/api/fs.html#fs_fs_exists_path_callback

use `stat` instead and here since there is no need for parallelism, you can even use the sync ones :)

Comment (nginnever) [inline tests/test-import.js]: ahh ty for pointing me to fs.stat, didn't realize exists was the deprecated version, will change for a sync version, also was checking twice on the level-2 to cover a weird case that someone would have dir-another but not level-2 or vice versa. Either way that should only be executed when the repo is cloned so it could be assumed both won't exist.

**Comment by diasdavid** [comment]
@nginnever what's missing on this one for it to be merged?

**Comment by nginnever** [comment]
@diasdavid I'm about to push some commits I worked on over the weekend after the help I got from @dignifiedquire 

So I have all the fixed-sized-chunker tests passing and will skip the web tests for file/dir importer and only test the buffer importer. I will get the buffer tests running today. I was thinking about adding the stream importer and tests in this PR but I still have a couple of questions about how we will flag the stream as a file or a dir or the best approach to this. So maybe I should make a separate PR for stream importer. 

After that I just need to standard the code up and have you take a look. 

Thanks!

File: karma.conf.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
+'use strict'
```
Comment (diasdavid) [inline karma.conf.js]: is this needed?

File: package.json [hunk 8]
```diff
@@ -5,9 +5,13 @@
   "main": "src/index.js",
   "scripts": {
     "lint": "standard",
-    "coverage": "istanbul cover --print both -- _mocha tests/index.js",
-    "test": "mocha tests/index.js"
+    "test": "mocha tests/index.js && npm run test:browser",
```
Comment (diasdavid) [inline package.json]: could you break it to  test:node and test:browser?

File: src/index.js [hunk 9]
```diff
@@ -7,7 +7,7 @@ const FixedSizeChunker = require('./chunker-fixed-size')
 const through2 = require('through2')
 const UnixFS = require('ipfs-unixfs')
 const async = require('async')
-
+var isNode = !global.window
```
Comment (diasdavid) [inline src/index.js]: why var?

File: tests/browser.js [hunk 10]
```diff
@@ -0,0 +1,195 @@
+/* globals describe, before, it */
```
Comment (diasdavid) [inline tests/browser.js]: `/* eslint-env mocha */`

File: tests/browser.js [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+      repoData.push({
+        key: key.replace('./', ''),
+        value: repoContext(key)
+      })
+    })
+
+    const mainBlob = store('ipfs')
+    const blocksBlob = store('ipfs/blocks')
+
+    async.eachSeries(repoData, (file, cb) => {
```
Comment (diasdavid) [inline tests/browser.js]: can you make a file to be the set up of things for the browser and then another for the tests? a la js-ipfs and friends :)

**Comment by diasdavid** [comment]
@nginnever CR'ed. Overall LGTM, I'll wait for travis to merge, meanwhile, I left some little nitpicks :)

File: src/index.js (hunk 9 above)
Comment (nginnever) [inline src/index.js]: actually we won't need to test the environment if we aren't going to use the file or dir importer in the browser. Will remove

Comment (nginnever) [inline src/index.js]: do we need some sort of error handling to return a message in case someone tries to use the file or dir importer in a browser context? not sure what is possible and what isn't

File: .travis.yml [hunk 12]
```diff
@@ -0,0 +1,22 @@
+
+language: node_js
+node_js:
+  - "4.0"
```
Comment (diasdavid) [inline .travis.yml]: let's do:
  - '4'
  - '5'

File: .travis.yml [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+
+branches:
+  only:
+    - master
```
Comment (diasdavid) [inline .travis.yml]: no need for this

File: package.json [hunk 14]
```diff
@@ -22,16 +27,33 @@
...
     "mocha": "^2.3.4",
     "ncp": "^2.0.0",
     "pre-commit": "^1.1.2",
+    "raw-loader": "^0.5.1",
     "rimraf": "^2.5.1",
-    "standard": "^5.4.1"
+    "standard": "^5.4.1",
+    "string-to-stream": "^1.0.1",
+    "transform-loader": "^0.2.3",
+    "webpack": "github:diasdavid/webpack#webpack-1"
```
Comment (diasdavid) [inline package.json]: Let's switch to webpack 2

File: src/index.js [hunk 15]
```diff
@@ -7,33 +7,48 @@ const FixedSizeChunker = require('./chunker-fixed-size')
...
 const UnixFS = require('ipfs-unixfs')
 const async = require('async')
-
 exports = module.exports
 
 const CHUNK_SIZE = 262144
 
 // Use a layout + chunkers to convert a directory (or file) to the layout format
-exports.import = (options, callback) => {
+exports.import = function (options, callback) {
```
Comment (diasdavid) [inline src/index.js]: why not an arrow function?

File: tests/test-import.js [hunk 16]
```diff
@@ -1,5 +1,4 @@
 /* globals describe, it */
```
Comment (diasdavid) [inline tests/test-import.js]: `/* eslint-env mocha */`

File: package.json [hunk 17]
```diff
@@ -22,16 +27,33 @@
...
+    "karma-sourcemap-loader": "^0.3.7",
+    "karma-spec-reporter": "0.0.24",
+    "karma-webpack": "^1.7.0",
     "mocha": "^2.3.4",
     "ncp": "^2.0.0",
     "pre-commit": "^1.1.2",
+    "raw-loader": "^0.5.1",
     "rimraf": "^2.5.1",
-    "standard": "^5.4.1"
+    "standard": "^5.4.1",
```
Comment (diasdavid) [inline package.json]: let's take the chance to upgrade to standard 6

File: src/index.js (hunk 15 above)
Comment (nginnever) [inline src/index.js]: arrow functions scare me after https://github.com/ipfs/js-ipfs-repo/blob/master/tests/browser-tests.js#L22

File: .travis.yml [hunk 18]
```diff
@@ -0,0 +1,19 @@
+
+language: node_js
+node_js:
+  - '4.0'
```
Comment (diasdavid) [inline .travis.yml]: just '4' and '5', so it picks the latest minor version :)
